Move crypto test suites

ronald-cron-arm opened the pull request:
## Description
Repo split preparation, move crypto test suites to tf-psa-crypto.
Companion mbedtls-framework PR: https://github.com/Mbed-TLS/mbedtls-framework/pull/36
Fix https://github.com/Mbed-TLS/mbedtls/issues/9266

## PR checklist
- [x] **changelog** not required because:  we will have one change log for the whole repo split work
- [x] **development PR** provided #9394
- [x] **framework PR** provided Mbed-TLS/mbedtls-framework#36
- [x] **3.6 PR** provided #9419 
- [x] **2.28 PR** not required because: split work
- **tests**  not required because: no additional test required

minosgalanakis commented on tests/scripts/basic-build-test.sh:
```diff
@@ -35,6 +35,8 @@ if [ -d library -a -d include -a -d tests ]; then :; else
     exit 1
 fi
 
+MBEDTLS_ROOT_DIR="$PWD"
```
This is a bit risky. It assumes that we are calling it from the actual root directory.

We could do something smart like asking git to tell us which is the root
`MBEDTLS_ROOT_DIR=$(git rev-parse --show-toplevel)`

But that won't work for tar-balls.

gilles-peskine-arm replied:
Many of our test scripts either change to the root or must be called from the root. `basic-build-test.sh` in particular must be called from the root (see the check near the top of the file).

ronald-cron-arm replied:
> This is a bit risky. It assumes that we are calling it from the actual root directory.

We have checked that we are calling it from the root directory just above.

minosgalanakis commented on tf-psa-crypto/tests/CMakeLists.txt:
```diff
@@ -0,0 +1,362 @@
+set(libs
```
This file is duplicating {MBEDTLS-ROOT}/tests/CMakeLists.txt.  It is effectively a large ammount of duplicated code. I am assuming those files are expected to diverge later on, so could we possibly source one from the other when they need to be identical?

ronald-cron-arm replied:
Eventually there will be one in mbedtls and one in TF-PSA-Crypto with quite similar contents and we do not plan to share anything there. Otherwise as of this PR, the two files diverge in the two following commits. I've done it that way to highlight in the commit history that the new tf-psa-crypto/tests/CMakeLists.txt file is just a derivation of the one we have in Mbed TLS and you can see the derivation in the commit "Add tf-psa-crypto test suites build".

ronald-cron-arm commented:
@minosgalanakis @davidhorstmann-arm I've rebased to resolve conflicts. While rebasing I had to manually resolve conflicts (path to framework directory) only in commit "Adapt path to framework data files". Otherwise I have added a new commit, last one, addressing the psasim issue and test_suite_with_psasim should run successfully now.

tom-cosgrove-arm commented on tests/psa-client-server/psasim/test/start_server.sh:
```diff
@@ -8,7 +8,7 @@ set -e
 # event as signal that the server is ready so that we can start client(s).
 function wait_for_server_startup() {
-    while [ $(find . -name "psa_notify_*" | wc -l) -eq 0 ]; do
+    while [ -z $(find /tmp -maxdepth 1 -name "psa_notify_*" -printf 1 -quit) ]; do
```
The use of `-printf` means that this will no longer work on macOS

gilles-peskine-arm replied:
Also putting files in `/tmp` is risky. We control build directories, so we should keep all files inside them.

I haven't read the whole code, but just looking at this line, it seems to me that if there are concurrent test runs on the same machine, they'll interfere with each other.

gilles-peskine-arm replied:
A portable way to check if there are `psa_notify_*` files in the current directory:

```
files_exist () {
    [ $# -ge 2 ] || [ -e "$1" ] || [ -L "$1" ]
}
while files_exist 'psa_notify_*'; do ...
```

ronald-cron-arm replied:
> The use of `-printf` means that this will no longer work on macOS

ah okay

 

> A portable way to check if there are `psa_notify_*` files in the current directory:
> 
> ```
> files_exist () {
>     [ $# -ge 2 ] || [ -e "$1" ] || [ -L "$1" ]
> }
> while files_exist 'psa_notify_*'; do ...
> ```

I can do that. @tom-cosgrove-arm okay with that proposal?

tom-cosgrove-arm replied:
And regarding concurrent @gilles-peskine-arm yes, this isn't great. Ideally the server would output it's doorbell path when run, and the client would use that - no "hoping" to get the right one. But this right here is a quick fix.

I agree `/tmp` could be problematic, the question would be: would it be better to just check for each of two possible directories the files could be created in within the build tree, and use the first one found?

tom-cosgrove-arm replied:
@ronald-cron-arm err, yes, I am okay with `files_exist`, but see my comment above - I do agree with @gilles-peskine-arm that `/tmp` has its problems

gilles-peskine-arm replied:
> would it be better to just check for each of two possible directories the files could be created in within the build tree, and use the first one found?

Sorry, I'm probably missing something because I haven't gone through the whole code and conversation. Why would we not know which directory the files are in?

tom-cosgrove-arm replied:
We used to put them in `./tests`, but that directory doesn't exist when we're in the framework (to a first approximation, I am multitasking severely ATM). So the suggestion: check if `./tests` is present. If so, use that, else use whatever directory is available when in the framework

ronald-cron-arm replied:
I will work on a better solution than /tmp.

ronald-cron-arm replied:
I've reverted the "Adapt psasim" commit and am proposing another solution to be able to run test suites located in `tests` and `tests/tf-psa-crypto`: starting two PSA servers, one "in" `tests`" and one "in" `tf-psa-crypto/tests`. @minosgalanakis @tom-cosgrove-arm please have a look, thanks.

minosgalanakis reviewed (commented):
Good overall, couple of minor tibits. 

Also as a side comment we spend a lot of code for directory discovery across multiple scripts.

e.g 

        if os.path.isdir(os.path.join(git_worktree_path, 'tf-psa-crypto',
                                      'tests', 'suites')):

We are trying to detect if the script is run from:
- mbedtls
- tf-psa-crypto
- framework

We already have this or similar detection logic implemented across different scripts

~~~~~~~
in_mbedtls_repo () {
    test -d include -a -d library -a -d programs -a -d tests
}

in_tf_psa_crypto_repo () {
    test -d include -a -d core -a -d drivers -a -d programs -a -d tests
}
~~~~~~~

Could we just put in a simple bash script, that all files, from bash, Makefiles, Cmake, and python use to test? 

This can come in an following PR this one is already too big.

ronald-cron-arm commented on tests/psa-client-server/psasim/Makefile:
```diff
@@ -77,5 +77,5 @@ clean_server_intermediate_files:
 	rm -f test/psa_client_base test/psa_client_full test/psa_server
 	rm -rf client_libs server_libs
-	rm -f test/psa_service_* test/psa_notify_* test/*.log
+	rm -f test/*.log
```
I've reverted the commit "Adapt psasim".

ronald-cron-arm commented:
> Also as a side comment we spend a lot of code for directory discovery across multiple scripts.
> 
> e.g
> 
> ```
>     if os.path.isdir(os.path.join(git_worktree_path, 'tf-psa-crypto',
>                                   'tests', 'suites')):
> ```
> 
> We are trying to detect if the script is run from:
> 
>     * mbedtls
> 
>     * tf-psa-crypto
> 
>     * framework
> 
> 
> We already have this or similar detection logic implemented across different scripts
> 
> ```
> in_mbedtls_repo () {
>     test -d include -a -d library -a -d programs -a -d tests
> }
> 
> in_tf_psa_crypto_repo () {
>     test -d include -a -d core -a -d drivers -a -d programs -a -d tests
> }
> ```
> 
> Could we just put in a simple bash script, that all files, from bash, Makefiles, Cmake, and python use to test?
> 
> This can come in an following PR this one is already too big.

Yes that's the plan and will be done when working on https://github.com/Mbed-TLS/mbedtls-framework/issues/16 and https://github.com/Mbed-TLS/mbedtls-framework/issues/12. We have already started some work around that, see #8609.

gilles-peskine-arm commented:
> Also as a side comment we spend a lot of code for directory discovery across multiple scripts.

In Python, please use the `build_tree` module.

tom-cosgrove-arm commented on tests/scripts/all.sh:
```diff
@@ -1098,16 +1098,24 @@ helper_psasim_server() {
         psa-client-server/psasim/test/start_server.sh
+        msg "start server in tf-psa-crypto/tests"
+        cd ../tf-psa-crypto/tests
+        ../../tests/psa-client-server/psasim/test/start_server.sh
```
Should we just do one of these, depending what directory/directories exist (`-d foo`)?

ronald-cron-arm replied:
No we need both. With this PR we now have x509 and TLS test suites in `tests` and the crypto test suites in `tf-psa-crypto/tests`. When in `test_suite_with_psasim` we call `make PSASIM=1 test` we first execute the test suites located in `tests` and for these ones we need a PSA server started from `tests` and then we run the test suites located in `tf-psa-crypto/tests` for which we need a PSA server started from `tf-psa-crypto/tests`.

tom-cosgrove-arm replied:
Ah okay, thanks

tom-cosgrove-arm commented:
The PSASIM changes look good to me

davidhorstmann-arm reviewed (changes requested):
A couple of small things, LGTM otherwise

ronald-cron-arm commented on tests/Makefile:
```diff
@@ -122,29 +126,35 @@ generated_psa_test_data: ../framework/scripts/mbedtls_framework/test_data_genera
+CRYPTO_DATA_FILES += $(GENERATED_CRYPTO_DATA_FILES)
+
 APPS = $(basename $(subst suites/,,$(DATA_FILES)))
+CRYPTO_APPS = $(basename $(subst suites/,,$(CRYPTO_DATA_FILES)))
```
I don't think so. We want the path to the crypto test suites executables in `tf-psa-crypto/tests/` and they are of the form: ../tf-psa-crypto/tests/test_suite.. Thus here we just remove the `suites/` part of the path to the *.data files.

davidhorstmann-arm replied:
Ah you're right, I was getting confused, thanks!

ronald-cron-arm commented on tests/CMakeLists.txt:
```diff
@@ -219,49 +105,16 @@ function(add_test_suite suite_name)
-            ${CMAKE_CURRENT_BINARY_DIR}/suites/test_suite_${data_name}.data)
-        set(dependency test_suite_bignum_generated_data)
-    elseif(";${config_generated_data_names};" MATCHES ";${data_name};")
+    if(";${config_generated_data_names};" MATCHES ";${data_name};")
```
The path to the data file is based on CMAKE_CURRENT_BINARY_DIR in one case, CMAKE_CURRENT_SOURCE_DIR otherwise thus that's one difference. Otherwise, the `set(dependency test_suite_config_generated_data)` can be moved out of the if-else.

davidhorstmann-arm replied:
Ah I did indeed miss something, thanks!

ronald-cron-arm commented on scripts/bump_version.sh:
```diff
@@ -124,8 +124,8 @@ cat include/mbedtls/build_info.h |                                    \
 [ $VERBOSE ] && echo "Bumping version in tests/suites/test_suite_version.data"
-sed -e "s/version:\".\{1,\}/version:\"$VERSION\"/g" < tests/suites/test_suite_version.data > tmp
-mv tmp tests/suites/test_suite_version.data
+sed -e "s/version:\".\{1,\}/version:\"$VERSION\"/g" < tf-psa-crypto/tests/suites/test_suite_version.data > tmp
```
I think we will eventually have one mbedtls_version.c in Mbed TLS, one tf_psa_crypto_version.c in tf-psa-crypto and one version test suite in both of them. This should be done as part of the work on generate_features.pl and generate_query_config.pl (https://github.com/Mbed-TLS/mbedtls-framework/issues/16).

ronald-cron-arm commented:
@minosgalanakis @davidhorstmann-arm I believe I have addressed your comments. Please have another look. Otherwise I've created the 3.6 very limited backport (mainly to update in 3.6 the framework as done in this PR): https://github.com/Mbed-TLS/mbedtls/pull/9419. Please review it as well, thanks.

minosgalanakis reviewed (dismissed):
LGTM

davidhorstmann-arm reviewed (approved):
LGTM, thanks!

davidhorstmann-arm reviewed (dismissed):
LGTM, thanks!

ronald-cron-arm commented:
@davidhorstmann-arm @minosgalanakis last version: updated the framework to the merge of https://github.com/Mbed-TLS/mbedtls/issues/38 instead of https://github.com/Mbed-TLS/mbedtls/pull/36 and updated the PSA test wrappers.

minosgalanakis reviewed (approved):
LGTM

davidhorstmann-arm reviewed (approved):
LGTM, thanks!

gilles-peskine-arm commented on tests/Makefile:
```diff
@@ -235,8 +266,9 @@ else
-check: $(BINARIES)
+check: $(BINARIES) $(CRYPTO_BINARIES)
 	perl scripts/run-test-suites.pl $(TEST_FLAGS) --skip=$(SKIP_TEST_SUITES)
+	cd ../tf-psa-crypto/tests && perl ../../tests/scripts/run-test-suites.pl $(TEST_FLAGS) --skip=$(SKIP_TEST_SUITES)
```
[Bug: the crypto tests are skipped if a TLS test fails](https://github.com/Mbed-TLS/mbedtls/pull/9557).

Why doesn't `run-test-suites.pl` take care of looking for the subproject's test suites? (Like `ctest` does.)

gilles-peskine-arm replied:
>  Why doesn't `run-test-suites.pl` take care of looking for the subproject's test suites?

https://github.com/Mbed-TLS/mbedtls/pull/9558
